modules/core/src/ocl.cpp

```diff
@@ -2925,6 +2925,44 @@ CV_EXPORTS bool useSVM(UMatUsageFlags usageFlags)
 #endif // HAVE_OPENCL_SVM
 
 
+void attachContext(String& platformName, void* platformID, void* context, void* deviceID)
+{
+    cl_platform_id pl;
+    cl_device_id   d;
+
+    d = (cl_device_id)deviceID;
+
+    // get platform ID from user supplied device
+    CV_OclDbgAssert(clGetDeviceInfo(d, CL_DEVICE_PLATFORM, sizeof(cl_platform_id), &pl, NULL) == CL_SUCCESS);
+
+    // get platform name string length
+    size_t sz;
+    CV_OclDbgAssert(clGetPlatformInfo(pl, CL_PLATFORM_NAME, 0, 0, &sz) == CL_SUCCESS);
+
+    // get platform name string
+    AutoBuffer<char> buf(sz+1);
+    CV_OclDbgAssert(clGetPlatformInfo(pl, CL_PLATFORM_NAME, sz, buf, 0) == CL_SUCCESS);
+    // just in case, ensure trailing zero for ASCIIZ string
+    buf[sz] = 0;
+
+    String actualPlatformName = (const char*)buf;
+
+    // can't continue if actualPlatform doesn't match user supplied platform
+    CV_OclDbgAssert(platformName == actualPlatformName);
+
+    // do not initialize context
+    Context ctx = Context::getDefault(false);
+
+    initializeContextFromHandle(ctx, platformID, context, deviceID);
+
+    // clear command queue, if any
+    getCoreTlsData().get()->oclQueue.finish();
+    Queue q;
+    getCoreTlsData().get()->oclQueue = q;
```

Comment:
I still believe that "cleanup" should be done before context replacement and for all active threads (but there are no api for this yet).